geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/pubsub/Publisher.java

```diff
@@ -0,0 +1,330 @@
+              .setArguments(batch)
+              .execute(PublishFunction.ID);
+        }
+      } finally {
+        // execute it locally
+        publishBatchToLocalSubscribers(batch);
+        if (resultCollector != null) {
+          // block until remote execute completes
+          resultCollector.getResult();
+        }
+      }
+    } catch (Exception e) {
+      // the onMembers contract is for execute to throw an exception
+      // if one of the members goes down.
+      if (logger.isDebugEnabled()) {
+        logger.debug(
+            "Exception executing publish function on batch {}. If a server departed during the publish then an exception is expected.",
+            batch, e);
+      }
+    }
+  }
+
+  private void publishBatchToLocalSubscribers(List<PublishRequest> batch) {
+    batch.forEach(request -> subscriptions.forEachSubscription(request.getChannel(),
+        (subscriptionName, channelToMatch, client, subscription) -> subscription.publishMessage(
+            channelToMatch != null, subscriptionName,
+            client, request.getChannel(), request.getMessage())));
+  }
+
+  public static class PublishRequest implements DataSerializableFixedID {
+    // note final can not be used because of DataSerializableFixedID
+    private byte[] channel;
+    private byte[] message;
+
+    public PublishRequest(byte[] channel, byte[] message) {
+      this.channel = channel;
+      this.message = message;
+    }
+
+    @SuppressWarnings("unused")
+    public PublishRequest() {
+      // needed for DataSerializableFixedID
+    }
+
+    public byte[] getChannel() {
+      return channel;
+    }
+
+    public byte[] getMessage() {
+      return message;
+    }
+
+    @Override
+    public int getDSFID() {
+      return PUBLISH_REQUEST;
+    }
+
+    @Override
+    public void toData(DataOutput out, SerializationContext context) throws IOException {
+      DataSerializer.writeByteArray(getChannel(), out);
+      DataSerializer.writeByteArray(getMessage(), out);
+    }
+
+    @Override
+    public void fromData(DataInput in, DeserializationContext context)
+        throws IOException, ClassNotFoundException {
+      channel = DataSerializer.readByteArray(in);
+      message = DataSerializer.readByteArray(in);
+    }
+
+    @Override
+    public KnownVersion[] getSerializationVersions() {
+      return null;
+    }
+
+    @Override
+    public String toString() {
+      return "PublishRequest{" +
+          "channel=" + bytesToString(channel) +
+          ", message=" + bytesToString(message) +
+          '}';
+    }
+  }
+
+  /**
+   * Manages all the publish requests from a particular client.
+   * The order of these requests is maintained, and they will be
+   * delivered to subscribers in that order.
+   * Since the delivery to subscribers can be slower than the rate
+   * at which publish requests arrive, this publisher supports
+   * batching so allow multiple requests to be sent with a single
```

Comment:
Typo here, I think, should "so allow" be "to allow"?

Reply:
done